mesonbuild/interpreter.py

```diff
@@ -3076,6 +3083,16 @@ def find_program_impl(self, args, for_machine: MachineChoice = MachineChoice.HOS
         if progobj is None and args[0].endswith('python3'):
             prog = dependencies.ExternalProgram('python3', mesonlib.python_command, silent=True)
             progobj = ExternalProgramHolder(prog)
+        if progobj is None and fallback is not None:
+            kwargs = {
+                'required': required,
+                'silent': silent,
+                'wanted': wanted,
+                'search_dirs': search_dirs,
+                'fallback': fallback,
+            }
+            progobj = self.dependency_fallback(args[0], kwargs)
```

Comment:
dependency_fallback() assumes the variable you take from the subproject is a dependency object. This should fail if you're trying to take an executable/program object.

Reply:
You'll need to write a function almost exactly the same, but with messages tweaked to speak about programs instead of dependency, and do not call self.get_subproject_dep() but instead just something like `self.subprojects[dirname].get_variable_method([varname], {})` and verify it is an Executable or ExternalProgramHolder.

Reply:
Sorry, saw below that you hacked get_subproject_dep() too. But you still need to differentiate both code paths, hopefully without duplicating too much code.

Reply:
Oh, and you need the program is found, but the version does not match (the case `if not is_found` below), you also need to try to fallback too.